### PR TITLE
Upload images #1062

### LINKED ISSUE #1062: Upload images
#### Tasks:
- [x] Use **Uppy** with **XHR** integration for the **upload dialog**.
- [x] Implement this dialog in the **common** components folder for reusability.

### PR DESCRIPTION
**Pull request by joshuadkitenge**
## Description
- Upload images dialog using uppy xhr
- add image editor 

## Testing instructions

Add a set up instructions describing how the reviewer should test the code

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage
- [ ] {more steps here}

## Agile board tracking
closes #1062

### CODE REVIEW

File: src/api/api.types.tsx [hunk 1]
```diff
@@ -232,8 +232,25 @@ export interface UploadInfo {
...
+  title?: string | null;
+  description?: string | null;
+  upload_file: File;
+}
+
+export interface Image
+  extends Required<Omit<ImagePost, 'upload_file'>>,
+    CreatedModifiedMixin {
+  id: string;
+  thumbnail: string;
```
Comment (joshuadkitenge) [inline src/api/api.types.tsx:255]:
```suggestion
  thumbnail_base64: string;
```

File: src/common/images/uploadImagesDialog.component.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+import '@uppy/core/dist/style.css';
+import '@uppy/dashboard/dist/style.css';
+import ProgressBar from '@uppy/progress-bar'; // Import the ProgressBar plugin
+import { Dashboard } from '@uppy/react';
+import XHR from '@uppy/xhr-upload';
+import React from 'react';
+import { settings } from '../../settings';
+
+const MAX_FILE_SIZE_MB = 500;
+const MAX_FILE_SIZE_B = MAX_FILE_SIZE_MB * 1048576;
```
Comment (joshuadkitenge) [inline src/common/images/uploadImagesDialog.component.tsx:20]:
```suggestion
const MAX_FILE_SIZE_B = MAX_FILE_SIZE_MB * 1000 * 1000;
```

File: src/common/actionMenu.component.test.tsx [hunk 3]
```diff
@@ -108,6 +109,28 @@ describe('ActionMenu Component', () => {
...
+    const closeButton = screen.getByRole('button', { name: 'Close Modal' });
+    await user.click(closeButton);
+    await waitFor(() => {
+      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
+    });
+  });
+  it('opens the upload images dialog', async () => {
+    createView();
+
+    // Open the menu
```
Comment (joshuadkitenge) [inline src/common/actionMenu.component.test.tsx:121]:
```suggestion
    // Open the menu
```

File: src/api/api.types.tsx [hunk 4]
```diff
@@ -238,3 +238,20 @@ export interface AttachmentPostMetadataResponse
...
+  description?: string | null;
+  upload_file: File;
+}
+
+export interface Image
+  extends Required<Omit<ImagePost, 'upload_file'>>,
+    CreatedModifiedMixin {
+  id: string;
+  thumbnail_base64: string;
+}
```
Comment (joshuadkitenge) [inline src/api/api.types.tsx:257]:
```suggestion
export interface Image
  extends Required<Omit<ImagePost, 'upload_file'>>,
    CreatedModifiedMixin {
  id: string;
  thumbnail_base64: string;
  primary:boolean;
}
```

File: src/mocks/handlers.ts [hunk 5]
```diff
@@ -975,4 +975,23 @@ export const handlers = [
...
+    return HttpResponse.json(
+      {
+        entity_id: '6707ff91d32eaac11c6d8f01',
+        title: null,
+        description: null,
+        created_time: '2024-10-12T20:51:47.687000Z',
+        modified_time: '2024-10-12T20:51:47.687000Z',
+        id: '670ae163532106fbaa41f348',
+        file_name: 'logo.png',
+        thumbnail:
```
Comment (joshuadkitenge) [inline src/mocks/handlers.ts:991]:
```suggestion
        thumbnail_base64:
```

File: src/mocks/handlers.ts [hunk 6]
```diff
@@ -975,4 +975,23 @@ export const handlers = [
...
+  http.post('/images', async () => {
+    return HttpResponse.json(
+      {
+        entity_id: '6707ff91d32eaac11c6d8f01',
+        title: null,
+        description: null,
+        created_time: '2024-10-12T20:51:47.687000Z',
+        modified_time: '2024-10-12T20:51:47.687000Z',
+        id: '670ae163532106fbaa41f348',
+        file_name: 'logo.png',
```
Comment (joshuadkitenge) [inline src/mocks/handlers.ts:990]:
```suggestion
        file_name: 'logo.png',
        primary: false;
```

File: cypress/e2e/with_mock_data/items.cy.ts [hunk 7]
```diff
@@ -604,6 +604,105 @@ describe('Items', () => {
...
+      cy.findBrowserMockedRequests({
+        method: 'POST',
+        url: '/images',
+      }).should(async (postRequests: Request[]) => {
+        expect(postRequests.length).eq(2);
+      });
+      cy.findByText('Complete').should('exist');
+    });
+
+    it('displays error if post is unsuccessfully', () => {
```
Comment (joelvdavies) [inline cypress/e2e/with_mock_data/items.cy.ts:658]:
```suggestion
    it('displays error if post is unsuccessful', () => {
```

File: src/common/images/uploadImagesDialog.component.test.tsx [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    await user.click(await screen.findByText('Save changes'));
+
+    await user.click(await screen.findByText('Upload 1 file'));
+
+    expect(xhrPostSpy).toHaveBeenCalledWith('POST', '/images', true);
+
+    expect(await screen.findByText('Complete')).toBeInTheDocument();
+  });
+
+  it('displays error if post is unsuccessfully', async () => {
```
Comment (joelvdavies) [inline src/common/images/uploadImagesDialog.component.test.tsx:112]:
```suggestion
  it('displays error if post is unsuccessful', async () => {
```

File: src/common/images/uploadImagesDialog.component.test.tsx [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    const dropZone = screen.getByText('files cannot be larger than', {
+      exact: false,
+    });
+
+    // Create a drag-and-drop event for the file
+    Object.defineProperty(dropZone, 'files', {
+      value: [file1],
+    });
+
+    // Fire the drop event
```
Comment (joelvdavies) [inline src/common/images/uploadImagesDialog.component.test.tsx:80]:
```suggestion
    // Fire the drop event
```
Could you remove these where they occur elsewhere too please (I think a few are in the attachments?). Same for other unnecessary ones.

File: src/common/images/uploadImagesDialog.component.tsx [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+      // Prepare the updated metadata object
+      const updatedFileData = {
+        ...restMeta,
+        ...(formattedTitle && { title: formattedTitle }),
+        ...(formattedDescription && { description: formattedDescription }),
+      };
+
+      // Use patchFilesState to update the metadata
+      // Ensure you call the method properly
```
Comment (joshuadkitenge) [inline src/common/images/uploadImagesDialog.component.tsx:81]:
```suggestion
```

File: src/common/actionMenu.component.tsx
[the remark this answers is not in the document]
Comment (joshuadkitenge) [reply src/common/actionMenu.component.tsx:1]: change order

**Review by joelvdavies — APPROVED** [review]
Looks good, works well.

**Pull request merged**